### ISSUE TITLE
Bad conversion quality after retraining

### ISSUE THREAD

**Issue by kvnsq** [auspicious3000/autovc#33]
Hi,
first of all thanks for the great work on the AutoVC system.
I have tried to replicate the system, but could not achieve to achieve nearly the same quality as the pre-trained system.
I use the the same pre-processing for the mel-spectrograms as discussed in issue #4
and have trained the system with he same 20 VCTK speakers of the experiment of the paper (additionally with 8 speakers from the VCC data set, however results were similar when they were omitted).
I additionally used one-hot encodings instead of speaker embeddings of an encoder.

I trained for about 300.000 steps using Adam with default parameters and a LR of 0.0001, the train loss is about 6.67e-3 and the validation loss is about 0.01 and rising.  I've also tried out other learning rates (0.001, 0.0005) with no improvement of quality. The converted mel-spectrograms are still blurry and produce a low-quality, robotic voice.
In comparison, the converted mel-spectrograms of the supplied autovc model are much sharper and produce more natural voice, even when used with Griffin-Lim.
Here are the mel-spectrograms of my retrained model and the model of the repo

Retrained model            |  Supplied model
:-------------------------:|:-------------------------:
![p270-p228-own](https://user-images.githubusercontent.com/58212792/69620007-e9834d80-103c-11ea-8d9c-4f704a167c88.png) |  ![p270-p228-paper](https://user-images.githubusercontent.com/58212792/69620016-edaf6b00-103c-11ea-81d4-f7cc510ab76f.png)

Here is a minimal example of the loss and training loop I use.
I can also provide more of my code if wanted.

```python
def train_step(mel_spec_batch, embeddings_batch, generator,
               weight_mu_zero_rec: float, weight_lambda_content: float):
    optimizer.zero_grad()

    mel_spec_batch_exp = mel_spec_batch.unsqueeze(1) # (batch_size=2, 1, num_frames=128, num_mels=80)
    mel_outputs, mel_outputs_postnet, content_codes_mel_input = generator(mel_spec_batch,
                                                                          embeddings_batch,
                                                                          embeddings_batch)
    # Returns content codes with self.encoder without using the decoder and postnet a second time
    content_codes_gen_output = generator.get_content_codes(mel_outputs_postnet, embeddings_batch)

    rec_loss = F.mse_loss(input=mel_outputs_postnet, target=mel_spec_batch_exp, reduction="mean")
    rec_0_loss = F.mse_loss(input=mel_outputs, target=mel_spec_batch_exp, reduction="mean")
    content_loss = F.l1_loss(input=content_codes_gen_output, target=content_codes_mel_input, reduction="mean")
    total_loss = rec_loss + weight_mu_zero_rec * rec_0_loss + weight_lambda_content * content_loss

    total_loss.backward()
    optimizer.step()


# Train loop..
for epoch in range(start_epoch + 1, args[FLAGS.MAX_NUM_EPOCHS] + 1):
    generator.train()
    # Iterate over Mel-Spec Slices and the index of their speakers
    for step_idx, (mel_spec_batch, speaker_idx_batch) in enumerate(train_set_loader):
        # Load the speaker embeddings of the speakers of the mel-spectograms
        spkr_embeddings = speaker_embedding_mat[speaker_idx_batch.to(device)].to(device)
	train_step(mel_spec_batch.to(device), spkr_embeddings, generator, optim,
                   weight_mu_zero_rec=args[FLAGS.AUTO_VC_MU_REC_LOSS_BEFORE_POSTNET], # == 1.0
                   weight_lambda_content=args[FLAGS.AUTO_VC_LAMBDA_CONTENT_LOSS]) # == 1.0
	# The rest is computing the validation loss, resynthesizing utterances, saving the model every n epochs, etc
```

Does anyone have an idea what is wrong with my re-implementation or could anyone reimplement the system with good quality?

Thanks a lot in advance.

**Comment by dodobyte** [comment]
I think nobody reproduced the same quality yet, you can check the other issues. I tried many hyperparameters and I couldn't produce mel any better than you. Here's an example.

![](https://i.imgur.com/pNjWiWx.png)

I used 128 frame 80 dimensional mels padded with zero (or trimmed for longer mels). Tried learning rate both 1e-3 and 1e-4.
Tried VCTK with 20 and 40 speakers. Tried with masked loss and MSE+L1 directly like yours. Tried 2 and 4 minibatch sizes.

Also I verified that my mels are almost exactly same with the example mels in the repo and works well with the pretrained model.

**Comment by auspicious3000** [comment]
@dodobyte Someone reproduced results and published in ICLR. Unfortunately, those who successfully produced the results don't post here.

**Comment by auspicious3000** [comment]
@kvnsq Is your dataloader randomized?

**Comment by kvnsq** [comment]
> @kvnsq Is your dataloader randomized?

No, it wasn't. Thanks for the tip, I will try it out. 
I've divided mel-specs to fixed 128 frame slices, shuffled them, and iterated over them.  Now I iterate through shuffled mel-specs and randomly crop them on the fly because of your suggestion.
Do you also randomly sample mel-specs from the train set in each training step, such that the same samples could be in two consecutive mini-batches?

In my previous tries I noticed that I could increase the batch size to 32  without a decrease in the loss compared to a batch size of 2.  Because training is much faster that way, I've rerun the experiments with the on-the-fly random cropping, but couldn't see an increase in quality. I'm going to retrain the model with a batch size of 2 and LR of  0.0001 overnight and report if I see any changes in quality.

**Comment by youngsuenXMLY** [comment]
> Hi,
> first of all thanks for the great work on the AutoVC system.
> I have tried to replicate the system, but could not achieve to achieve nearly the same quality as the pre-trained system.
> I use the the same pre-processing for the mel-spectrograms as discussed in issue #4
> and have trained the system with he same 20 VCTK speakers of the experiment of the paper (additionally with 8 speakers from the VCC data set, however results were similar when they were omitted).
> I additionally used one-hot encodings instead of speaker embeddings of an encoder.
> 
> I trained for about 300.000 steps using Adam with default parameters and a LR of 0.0001, the train loss is about 6.67e-3 and the validation loss is about 0.01 and rising. I've also tried out other learning rates (0.001, 0.0005) with no improvement of quality. The converted mel-spectrograms are still blurry and produce a low-quality, robotic voice.
> In comparison, the converted mel-spectrograms of the supplied autovc model are much sharper and produce more natural voice, even when used with Griffin-Lim.
> Here are the mel-spectrograms of my retrained model and the model of the repo
> 
> Retrained model	Supplied model
> ![p270-p228-own](https://user-images.githubusercontent.com/58212792/69620007-e9834d80-103c-11ea-8d9c-4f704a167c88.png)	![p270-p228-paper](https://user-images.githubusercontent.com/58212792/69620016-edaf6b00-103c-11ea-81d4-f7cc510ab76f.png)
> Here is a minimal example of the loss and training loop I use.
> I can also provide more of my code if wanted.
> 
> ```python
> def train_step(mel_spec_batch, embeddings_batch, generator,
>                weight_mu_zero_rec: float, weight_lambda_content: float):
>     optimizer.zero_grad()
> 
>     mel_spec_batch_exp = mel_spec_batch.unsqueeze(1) # (batch_size=2, 1, num_frames=128, num_mels=80)
>     mel_outputs, mel_outputs_postnet, content_codes_mel_input = generator(mel_spec_batch,
>                                                                           embeddings_batch,
>                                                                           embeddings_batch)
>     # Returns content codes with self.encoder without using the decoder and postnet a second time
>     content_codes_gen_output = generator.get_content_codes(mel_outputs_postnet, embeddings_batch)
> 
>     rec_loss = F.mse_loss(input=mel_outputs_postnet, target=mel_spec_batch_exp, reduction="mean")
>     rec_0_loss = F.mse_loss(input=mel_outputs, target=mel_spec_batch_exp, reduction="mean")
>     content_loss = F.l1_loss(input=content_codes_gen_output, target=content_codes_mel_input, reduction="mean")
>     total_loss = rec_loss + weight_mu_zero_rec * rec_0_loss + weight_lambda_content * content_loss
> 
>     total_loss.backward()
>     optimizer.step()
> 
> 
> # Train loop..
> for epoch in range(start_epoch + 1, args[FLAGS.MAX_NUM_EPOCHS] + 1):
>     generator.train()
>     # Iterate over Mel-Spec Slices and the index of their speakers
>     for step_idx, (mel_spec_batch, speaker_idx_batch) in enumerate(train_set_loader):
>         # Load the speaker embeddings of the speakers of the mel-spectograms
>         spkr_embeddings = speaker_embedding_mat[speaker_idx_batch.to(device)].to(device)
> 	train_step(mel_spec_batch.to(device), spkr_embeddings, generator, optim,
>                    weight_mu_zero_rec=args[FLAGS.AUTO_VC_MU_REC_LOSS_BEFORE_POSTNET], # == 1.0
>                    weight_lambda_content=args[FLAGS.AUTO_VC_LAMBDA_CONTENT_LOSS]) # == 1.0
> 	# The rest is computing the validation loss, resynthesizing utterances, saving the model every n epochs, etc
> ```
> 
> Does anyone have an idea what is wrong with my re-implementation or could anyone reimplement the system with good quality?
> 
> Thanks a lot in advance.

hi, I didn't find code about style encoder. Did you implement the style encoder?

**Comment by auspicious3000** [comment]
Please refer to #24, or simply use one-hot embeddings.

**Comment by zonerby** [comment]
@kvnsq Did you manage to improve training of model ? I have the same problem. Retrained model produce poor quality mel-spectrograms. I've tried 20/40 VCTK speakers, 2-32 batch size, various mel-specs generating frameworks.

**Comment by kvnsq** [comment]
@zonerby No, I gave up on this model and am currently working on other VC models. I might try to retrain the AutoVC model with other NN architectures in the future, but that has lower priority for me right now. If I do retrain it and see some improvements, I'll let you know.

**Comment by ye2020** [comment]
I finally get this model works on a [Chinese corpus](http://openslr.org/38/) and get quite good output quality even convert from unseen speaker to unseen speaker, I use 120 speakers from the corpus, 120 utterances per speaker, 1e-4 learning rate, 4 batch size, the content embedding is downsampled
 by size of 16 (i.e. `Generator(32, 256, 512, 16)`), the default size 32 not works, and I use a GE2E speaker encoder trained on a dataset combined by several Chinese corpus with total ~2800 speakers, I use `l1_loss` instead of `mse_loss` for all three losses, the model is trained for 370k steps and achieve ~0.045 training loss.

This is how one of the training sample looks like, I also plot the downsampled content embedding along with the upsampled one as well:
<img width="1439" alt="Screen Shot 2020-01-22 at 4 47 53 AM" src="https://user-images.githubusercontent.com/59328293/72842936-7ca2d180-3cd4-11ea-859c-922d6e167002.png">

**Comment by auspicious3000** [comment]
Please send me an email if any of you need the complete reproducible training code.

**Comment by PTaati** [comment]
If possible, please send the code to ventotaasai@gmail.com I promise not to be commercialized.

**Comment by LG-SS** [comment]
> I finally get this model works on a [Chinese corpus](http://openslr.org/38/) and get quite good output quality even convert from unseen speaker to unseen speaker, I use 120 speakers from the corpus, 120 utterances per speaker, 1e-4 learning rate, 4 batch size, the content embedding is downsampled
> by size of 16 (i.e. `Generator(32, 256, 512, 16)`), the default size 32 not works, and I use a GE2E speaker encoder trained on a dataset combined by several Chinese corpus with total ~2800 speakers, I use `l1_loss` instead of `mse_loss` for all three losses, the model is trained for 370k steps and achieve ~0.045 training loss.
> 
> This is how one of the training sample looks like, I also plot the downsampled content embedding along with the upsampled one as well:
> <img alt="Screen Shot 2020-01-22 at 4 47 53 AM" width="1439" src="https://user-images.githubusercontent.com/59328293/72842936-7ca2d180-3cd4-11ea-859c-922d6e167002.png">
> 
> This is a plot convert from unseen male to unseen female, I plot the source utterance, converted utterance, one of the target speaker utterance (the target speaker embedding is averaged from 5 utterances while inferencing), and the self-reconstruct utterance:
> <img alt="Screen Shot 2020-01-22 at 5 12 16 AM" width="746" src="https://user-images.githubusercontent.com/59328293/72843601-d061ea80-3cd5-11ea-8b0f-ad56e02acb42.png">

So can you upload some samples publically to judge the quality of the converted audios in chinese dataset? Thanks

**Comment by ye2020** [comment]
@LG-SS Try it yourself: https://apps.apple.com/cn/app/%E9%AD%94%E6%9C%AF%E5%8F%98%E5%A3%B0%E5%99%A8/id1499265894

**Comment by auspicious3000** [comment]
@ye2020 May I ask what vocoder did you use to convert from spectrograms to waveforms? Thx!

**Comment by ye2020** [comment]
@auspicious3000 It's melgan https://github.com/descriptinc/melgan-neurips/issues/4#issuecomment-570106464

**Comment by MayukhSobo** [comment]
@ye2020 Can you please please send me code. I have been struggling for a long time. Please email me on mayukh2012@hotmail.com. Please I really need help.

**Comment by shoutong** [comment]
> I finally get this model works on a [Chinese corpus](http://openslr.org/38/) and get quite good output quality even convert from unseen speaker to unseen speaker, I use 120 speakers from the corpus, 120 utterances per speaker, 1e-4 learning rate, 4 batch size, the content embedding is downsampled
> by size of 16 (i.e. `Generator(32, 256, 512, 16)`), the default size 32 not works, and I use a GE2E speaker encoder trained on a dataset combined by several Chinese corpus with total ~2800 speakers, I use `l1_loss` instead of `mse_loss` for all three losses, the model is trained for 370k steps and achieve ~0.045 training loss.
> 
> This is how one of the training sample looks like, I also plot the downsampled content embedding along with the upsampled one as well:
> <img alt="Screen Shot 2020-01-22 at 4 47 53 AM" width="1439" src="https://user-images.githubusercontent.com/59328293/72842936-7ca2d180-3cd4-11ea-859c-922d6e167002.png">
> 
> This is a plot convert from unseen male to unseen female, I plot the source utterance, converted utterance, one of the target speaker utterance (the target speaker embedding is averaged from 5 utterances while inferencing), and the self-reconstruct utterance:
> <img alt="Screen Shot 2020-01-22 at 5 12 16 AM" width="746" src="https://user-images.githubusercontent.com/59328293/72843601-d061ea80-3cd5-11ea-8b0f-ad56e02acb42.png">

Would you please publicly release your code and model. I've tried many times with different corpora but all results are merely nonsense noise.

**Comment by DouYishun** [comment]
Hi @himajin2045 Could you please explain why did you choose L1 loss instead of L2 in origin design?

**Comment by himajin2045** [comment]
@DouYishun It's been a year and I don't quite remember the exact reason, but I'm sure it was not a theoretical one.

**Comment by himajin2045** [comment]
@shoutong Sorry for the late reply, and for those who are interested in my code, here you go https://github.com/himajin2045/voice-conversion

**Comment by JohnHerry** [comment]
@himajin2045 Thanks for your good job. I am training a mandarin vc model now.  I am using the AISHell-3 dataset which is more then 70 hours speech with 174 different speakers.  The speaker embedding model I am using is the pretrained one in this project. And no parameter changed throuth my experiment.

But my vc model can product only noise,  no speech waveform is generated.  The reasons I guessed are as follows :

1、 the speaker embedding should be a new one trained with Mandarin corpus [I did not found a pretrained one with Mandarin]

2、Maybe the training corpus is too small, more samples needed.

3、There are same hparams should be changed before my training preprocessing.

4、The `make_spect.py` may need some change, for more commonly `mel` used in training process.

But I do not have clear sense which should I do.

Have you any suggestions? Thanks

**Comment by himajin2045** [comment]
@JohnHerry

Sorry, I didn't try the updated pretrained models published in this repo. But as you can see in  my previous comments, with 120 speakers is enough in my experience.

Perhaps you can try my code to train your own speaker encoder and vocoder.

**Comment by JohnHerry** [comment]
> @JohnHerry
> 
> Sorry, I didn't try the updated pretrained models published in this repo. But as you can see in my previous comments, with 120 speakers is enough in my experience.
> 
> Perhaps you can try my code to train your own speaker encoder and vocoder.

Thanks for your help, and I have noticed your project on VC, I will have a try.

**Comment by taubaaron** [comment]
> The complete reproducible training code may be sent through email upon special request for non-commercial purposes.

I would very much appreciate that if its okay with you.
aaronadiella@gmail.com

**Comment by auspicious3000** [comment]
@taubaaron Please disregard that message. The code has already been released.
